File: 714. Best Time to Buy and Sell Stock with Transaction Fee/main.cpp
#include <numeric>
#include <vector>

using namespace std;

class Solution {
public:
    int maxProfit(vector<int> &prices, int fee) {
        int buy = numeric_limits<int>::min();
        int profit = 0;

        for (const auto &price : prices) {
            buy = max(buy, profit - price);
            profit = max(profit, price + buy - fee);
        }

        return profit;
    }
};

File: 714. Best Time to Buy and Sell Stock with Transaction Fee/main.rs
struct Solution;

impl Solution {
    pub fn max_profit(prices: Vec<i32>, fee: i32) -> i32 {
        let mut buy = i32::MIN;
        let mut profit = 0;

        for price in prices {
            buy = buy.max(profit - price);
            profit = profit.max(buy + price - fee);
        }

        profit
    }
}
